### PR TITLE
Throw DNSError in getaddrinfo (hide UVError)

### PR DESCRIPTION
**Pull request by samoconnor**
See #14972 and #7841 and https://github.com/JuliaLang/julia/issues/6825#issuecomment-42912316

### CODE REVIEW

File: base/socket.jl [hunk 1]
```diff
@@ -582,13 +582,27 @@ function getaddrinfo(cb::Function, host::ASCIIString)
...
+
 function getaddrinfo(host::ASCIIString)
     c = Condition()
     getaddrinfo(host) do IP
         notify(c,IP)
     end
     ip = wait(c)
-    isa(ip,UVError) && throw(ip)
+    if isa(ip,UVError)
+        @assert ip.code in [UV_EAI_NONAME, UV_EAI_AGAIN, UV_EAI_FAIL, UV_EAI_NODATA]
```
Comment (tkelman) [inline base/socket.jl]: Again, an assert is not appropriate here. If the code is not in this list, rethrow it so the user can see what the code actually was.

**Comment by samoconnor** [comment]
> i would go with it. i'm still undecided on whether to change it to use direct return of a nullable error type. in the meantime, providing clearer error messages i think is still a worthwhile investment.

@vtjnash any interest in merging this?

**Comment by vtjnash** [comment]
yes, just waiting to see Tony's comment addressed (remove the assert and rethrow the UVError if it wasn't one of the expected ones)

File: base/socket.jl [hunk 2]
```diff
@@ -553,7 +563,7 @@ function uv_getaddrinfocb(req::Ptr{Void}, status::Cint, addrinfo::Ptr{Void})
     cb = unsafe_pointer_to_objref(data)::Function
     pop!(callback_dict,cb) # using pop forces an error if cb not in callback_dict
     if status != 0 || addrinfo == C_NULL
-        cb(UVError("getaddrinfo callback",status))
+        cb(UVError("uv_getaddrinfodb received an unexpected status code", status))
```
Comment (tkelman) [inline base/socket.jl]: uv_getaddrinfocb

**Comment by samoconnor** [comment]
@vtjnash OK, done.
Based on my reading of the libuv documentation and source code (win and unix), UVError will now only be thrown if there is a bug.

File: base/socket.jl [hunk 3]
```diff
@@ -577,8 +587,16 @@ end
 
 function getaddrinfo(cb::Function, host::ASCIIString)
     callback_dict[cb] = cb
-    uv_error("getaddrinfo",ccall(:jl_getaddrinfo, Int32, (Ptr{Void}, Cstring, Ptr{UInt8}, Any, Ptr{Void}),
-                                 eventloop(), host, C_NULL, cb, uv_jl_getaddrinfocb::Ptr{Void}))
+    status = ccall(:jl_getaddrinfo, Int32, (Ptr{Void}, Cstring, Ptr{UInt8}, Any, Ptr{Void}),
+                   eventloop(), host, C_NULL, cb, uv_jl_getaddrinfocb::Ptr{Void})
+    if status == UV_EINVAL
+        throw(ArgumentError("Invalid uv_getaddrinfo() agument"))
+    elseif status in [UV_ENOMEM, UV_ENOBUFS]
```
Comment (tkelman) [inline base/socket.jl]: ENOBUFS isn't really an OutOfMemoryError

Comment (samoconnor) [inline base/socket.jl]: It is [in this case](https://github.com/libuv/libuv/blob/v1.x/src/win/getaddrinfo.c#L305).

File: base/exports.jl [hunk 4]
```diff
@@ -171,6 +171,7 @@ export
     TypeError,
     AssertionError,
     UnicodeError,
+    DNSError,
```
Comment (tkelman) [inline base/exports.jl]: if it's going to be exported, it needs docs

Comment (samoconnor) [inline base/exports.jl]: Fair enough. I've un-exported for now.

**Comment by samoconnor** [comment]
squashed

File: base/exports.jl (hunk 4 above)
Comment (nalimilan) [inline base/exports.jl]: If it's thrown, then it should be exported, right?

Comment (tkelman) [inline base/exports.jl]: we don't export UVError

**Comment by samoconnor** [comment]
@vtjnash bump

**Pull request merged**